### ISSUE TITLE
Raise Error when calling `nn.Parameter` on already `nn.Parameter` instance.

### ISSUE THREAD

**Issue by PaulScemama** [pytorch/pytorch#93941]
### 🚀 The feature, motivation and pitch

I recently ran into a silent bug while implementing a simple bayesian neural network.

## Background
I created a Multivariate Gaussian distribution that takes in parameters `mu` and `rho`. The `__init__` then sets these as `nn.Parameters` because they are the parameters we need to optimize.

Afterwards I created a `BayesianLinearLayer` that created

```python
self.weight_mu = nn.Parameter(...)
self.weight_rho = nn.Parameter(...)
self.weight = multivariate_gaussian(mu=self.weight_mu, rho=self.weight_rho)

self.bias_mu = nn.Parameter(...)
self.bias_rho = nn.Paramter(...)
self.bias = multivariate_gaussian(mu=self.bias_mu, rho=self.bias_rho)
```

Because the `multivariate_gaussian` takes in ALREADY `NN.PARAMETER`s and then sets those again as `nn.Parameters`, taking any gradient wrt to these parameters was `None`. 

I wanted to raise this and see if this was thought to be (perhaps) a common mistake people make. I see now that it was my lack of pytorch knowledge that I made it, so maybe it is rare.

Feel free to close if you think it's such a rarity that it's not worth a change, but I feel like an error raised when trying to cast an `nn.Parameter` as an `nn.Parameter` may help? 

Thanks!

### Alternatives

Ignore since this doesn't come up often enough.

cc @albanD @mruberry @jbschlosser @walterddr @saketh-are

**Comment by jbschlosser** [comment]
Hey @PaulScemama, thanks for the report! I don't think we can change the behavior to error for this case, as this is BC-breaking and I'm aware of instances that rely on the current behavior.

**Comment by PaulScemama** [comment]
Hi @jbschlosser, that is not a worry! I thought there might be a good reason why it didn't throw an error. Thanks for the reply!
